feat: add createRefund GraphQL mutation

kieckhafer opened the pull request:
Part of #5158 
Impact: **minor**  
Type: **feature**

A new `createRefund` mutation allows you to create a refund for a particular payment on an order. This was previously a meteor only method. The meteor method will be removed with #5158.

## Breaking changes
None

## Testing
1. Place an order
1. Use the following mutation to create a refund on said order
1. Try to refund more than the charge was, and see that only the remaining amout was refunded

```
createRefund(
  amount: {AMOUNT}!,
  orderId: {ORDER-ID}!,
  paymentId: {PAYMENT-ID}!,
  reason: {Reason}) {
    order {
      {FIELDS-TO-RETREIVE}
    }
  }
```

`amount`, `orderId`, and `paymentId` are required. `reason` is optional. It's set up as a `String` to be used with multiple payment providers, however for `Stripe`, which is our only provider at the moment, they only accept 3 values for reason: `duplicate`, `fraudulent`, or `requested_by_customer`, so be sure to use one of those values if testing with Stripe. This limit will be built into the Operator UI if Stripe is the payment method. From their docs:

> String indicating the reason for the refund. If set, possible values are duplicate, fraudulent, and requested_by_customer. If you believe the charge to be fraudulent, specifying fraudulent as the reason will add the associated card and email to your block lists, and will also help us improve our fraud detection algorithms.


I've added tests here to test the permission / requirements of variables, but did not add tests to check a result, as the payment method has it's own tests for this element: https://github.com/reactioncommerce/reaction/blob/feat-kieckhafer-listRefundsGQLQuery/imports/plugins/included/payments-stripe/server/no-meteor/util/stripeListRefunds.test.js#L10

focusaurus commented on imports/plugins/core/orders/server/no-meteor/mutations/createRefund.js:
```diff
@@ -0,0 +1,111 @@
+import sendOrderEmail from "../util/sendOrderEmail";
+
+const inputSchema = new SimpleSchema({
+  amount: Number,
```
Does it make sense to set a minimum of 0 for amount?

kieckhafer replied:
Maybe not `0`, but maybe `0.01`, if it's possible to do that. I can't think of a circumstance where there is s `0` refund.

focusaurus commented on imports/plugins/core/orders/server/no-meteor/mutations/createRefund.test.js:
```diff
@@ -0,0 +1,119 @@
+    reason: "Customer was unsatisfied with purchase"
+  })).rejects.toThrowErrorMatchingSnapshot();
+});
+
```
Maybe test negative amount here.

kieckhafer commented:
@focusaurus I've added a `min` check at `0.01`, not 0. I can't think of a scenario where there would be a refund of `0`. If you think it should be `0` instead, I can change it to that.

Also added a test for this.

Ready for another look.

focusaurus commented on imports/plugins/core/orders/server/no-meteor/mutations/createRefund.js:
```diff
@@ -0,0 +1,114 @@
+const inputSchema = new SimpleSchema({
+  amount: {
+    type: Number,
+    min: 0.01
```
Talked with Eric D about this. Thinking `min: 0, exclusiveMin: true` is probably safer since some currencies might have small fractional positive values (Bitcoin!).

focusaurus commented on imports/plugins/core/orders/server/no-meteor/mutations/createRefund.js:
```diff
@@ -0,0 +1,114 @@
+  });
+
+  // Send email to notify customer of a refund
+  sendOrderEmail(context, order, "refunded");
```
I think maybe we should pass `updatedOrder` to this function. The old code passed `order` too but seems maybe a bug.

focusaurus commented on imports/plugins/core/orders/server/no-meteor/mutations/createRefund.test.js:
```diff
@@ -0,0 +1,132 @@
+    paymentId: "payment3",
+    reason: "Customer was unsatisfied with purchase"
+  })).rejects.toThrowErrorMatchingSnapshot();
+});
```
Can we add a test for a base success case? Looks like all these tests are error cases.

kieckhafer replied:
Test added

focusaurus commented on imports/plugins/included/marketplace/server/no-meteor/util/stripeCreateRefund.js:
```diff
@@ -10,15 +10,16 @@ import formatForStripe from "./formatForStripe";
  * @param {Object} context an object containing the per-request state
  * @param {Object} paymentMethod object containing transaction ID
  * @param {Number} amount the amount to be refunded
+ * @param {String} [reason] the reason for the refund
```
Can we link to the stripe docs describing the allowed values for `reason` here or put in a cheat sheet?

kieckhafer replied:
Yes... I also have an explanation in the comments in the client side code ( #5158 )

focusaurus commented on imports/plugins/included/payments-stripe/server/no-meteor/util/stripeCreateRefund.js:
```diff
@@ -10,16 +10,17 @@ import formatForStripe from "./formatForStripe";
  * @param {Object} context an object containing the per-request state
  * @param {Object} paymentMethod object containing transaction ID
  * @param {Number} amount the amount to be refunded
+ * @param {Number} [reason] the reason for the refund
```
(Repeat) Can we link to the stripe docs describing the allowed values for `reason` here or put in a cheat sheet?

kieckhafer replied:
Yes... I also have an explanation in the comments in the client side code ( #5158 )

kieckhafer commented:
@focusaurus all comments addressed, ready for another look
